[Feat] NetworkService 구조 Async/Await으로 변환, DTO 및 PerformanceMonitor, LoggingMonitor 구현

youz2me opened the pull request:
<!--
Prefix [#이슈번호] 작업 설명
예시 : Feat [#33] 마이페이지 뷰 구현
-->

## 📄 작업 내용
- Combine 기반으로 구현되어 있던 NetworkService를 Swift 6의 async/await으로 전환해 비동기 처리 방식을 개선했어요.
- 에러 타입 안정성을 위해 Swift 6의 typed throws를 적용했어요.
- Moya의 LoggingPlugin을 차용해 LoggingMonitor를 구현했어요.
- 태스크 성능 측정 및 성능 병목 지점 파악을 위한 PerformanceMonitor를 구현했어요. 
- 4차 스프린트 서버 명세에 따라 31개의 DTO를 구현했어요.

## 💻 주요 코드 설명

### NetworkService Combine → Async/Await 전환
- 기존 Combine의 `AnyPublisher` 기반 비동기 처리를 Swift 6의 async/await으로 전환했습니다.
- Alamofire의 `serializingData().value` API를 활용하여 간결하고 직관적인 비동기 처리를 구현했습니다.

**개선 전 (Combine)**
```swift
public protocol NetworkServiceProtocol {
    func request<T: Decodable>(_ endPoint: NetworkEndpoint) -> AnyPublisher<T, NetworkError>
}

public func request<T: Decodable>(_ endPoint: NetworkEndpoint) -> AnyPublisher<T, NetworkError> {
    // ... 요청 생성
    return dataRequest
        .publishData()
        .tryMap { response -> Data in
            // ... 처리
        }
        .decode(type: T.self, decoder: JSONDecoder())
        .mapError { error in
            errorMapper.map(error)
        }
        .eraseToAnyPublisher()
}
```

**개선 후 (Async/Await)**
```swift
public protocol NetworkServiceProtocol {
    func request<T: Decodable>(_ endPoint: NetworkEndpoint) async throws(NetworkError) -> T
}

public func request<T: Decodable>(_ endPoint: NetworkEndpoint) async throws(NetworkError) -> T {
    // ... 요청 생성
    do {
        let data = try await dataRequest.serializingData().value

        guard let httpResponse = dataRequest.response else {
            throw NetworkError.invalidResponse
        }

        return try await handleResponse(data: data, response: httpResponse)
    } catch let error as NetworkError {
        throw error
    } catch {
        throw errorMapper.map(error)
    }
}
```

### Swift 6 Typed Throws 적용
- 기존의 일반 `throws`를 `throws(NetworkError)`로 변경하여 컴파일 타임에 에러 타입을 명확히 했습니다.
- 호출부에서 에러 타입을 확신할 수 있고, 컴파일 타임에 에러 타입 체크가 가능하도록 구현해 타입 안정성을 높였습니다.
- catch 블록 없이도 특정 에러 타입 보장이 가능합니다.

```swift
// NetworkServiceProtocol
func request<T: Decodable>(_ endPoint: NetworkEndpoint) async throws(NetworkError) -> T

// ResponseHandlerProtocol
func handle<T: Decodable>(data: Data, response: HTTPURLResponse) async throws(NetworkError) -> T

// ErrorMapperProtocol
func map(_ error: Error) -> NetworkError
```

### 태스크 성능 측정 및 성능 병목 지점 파악을 위한 PerformanceMonitor 구현
- 처음에는 네트워크 통신 속도를 측정하기 위한 PerformancePlugin을 구현했는데, 성능 개선에 활용하면 좋을 것 같아 모든 작업의 성능을 측정할 수 있는 PerformanceMonitor를 구현했습니다.
- 동기/비동기 작업 모두 지원하며, 시간과 메모리 사용량을 측정합니다.
- @discardableResult를 사용해 측정 결과를 무시하고 작업 결과만 사용 가능하며, 반환값을 강제하지 않도록 구현했습니다.
- rethrows 키워드를 사용해 호출부에서 유연하게 에러 처리가 가능하도록 구현했습니다.
- 동기/비동기 작업을 분리해 어떤 작업도 모두 성능 측정이 가능하도록 구현했습니다.

public final class PerformanceMonitor {
    private var startTimes: [String: Date] = [:]
    private var startMemories: [String: UInt64] = [:]
    private let queue = DispatchQueue(label: "com.livith.performance-monitor")

    public init() {}

    // 동기 작업 측정
    @discardableResult
    public func measure<T>(
        _ taskName: String,
        operation: () throws -> T
    ) rethrows -> T {
        let startMemory = getMemoryUsage()
        let startTime = Date()

        let result = try operation()

        let endTime = Date()
        let endMemory = getMemoryUsage()
        let duration = endTime.timeIntervalSince(startTime)
        let memoryUsed = endMemory - startMemory

        printMetric(taskName: taskName, duration: duration, memoryUsed: memoryUsed)

        return result
    }

    // 비동기 작업 측정
    @discardableResult
    public func measureAsync<T>(
        _ taskName: String,
        operation: () async throws -> T
    ) async rethrows -> T {
        let startMemory = getMemoryUsage()
        let startTime = Date()

        let result = try await operation()

        let endTime = Date()
        let endMemory = getMemoryUsage()
        let duration = endTime.timeIntervalSince(startTime)
        let memoryUsed = endMemory - startMemory

        printMetric(taskName: taskName, duration: duration, memoryUsed: memoryUsed)

        return result
    }

    // 수동 측정 시작
    public func start(_ taskName: String) {
        queue.sync {
            startTimes[taskName] = Date()
            startMemories[taskName] = getMemoryUsage()
        }
    }

    // 수동 측정 종료
    public func end(_ taskName: String) {
        queue.sync {
            guard let startTime = startTimes[taskName],
                  let startMemory = startMemories[taskName] else {
                return
            }

            let endTime = Date()
            let endMemory = getMemoryUsage()
            let duration = endTime.timeIntervalSince(startTime)
            let memoryUsed = endMemory - startMemory

            printMetric(taskName: taskName, duration: duration, memoryUsed: memoryUsed)

            startTimes.removeValue(forKey: taskName)
            startMemories.removeValue(forKey: taskName)
        }
    }
```

**사용 예시:**
```swift
// 동기 작업 측정
let monitor = PerformanceMonitor()
let result = monitor.measure("Heavy Calculation") {
    // 무거운 연산 수행
    return calculateSomething()
}

// 비동기 작업 측정
let data = await monitor.measureAsync("Network Request") {
    try await networkService.request(endpoint)
}

// 수동 측정
monitor.start("Custom Task")
// ... 작업 수행
monitor.end("Custom Task")
```

### ResponseHandler 책임 명확화
- 초기에는 ResponseHandler가 HTTP 검증만 담당하고 디코딩은 NetworkService가 수행하도록 설계했지만 이는 책임 분리 원칙에 어긋난다고 생각했습니다.
- ResponseHandler는 응답을 처리하는 컴포넌트이고, HTTP 검증과 디코딩은 모두 응답 처리의 일부이기 때문에 ResponseHandler의 영역이라고 생각했습니다.
- ResponseHandler가 HTTP 검증과 디코딩을 모두 수행해 NetworkService는 네트워크 요청 생성 및 실행에만 집중하도록 개선해 단일 책임 원칙을 준수했습니다.

**개선 전 (잘못된 책임 분리)**
```swift
// ResponseHandler: HTTP 검증만
public func handleAsync(data: Data, response: HTTPURLResponse) async throws(NetworkError) -> Data {
    let statusCode = response.statusCode
    guard (200..<300).contains(statusCode) else {
        throw NetworkError.from(statusCode: statusCode)
    }
    return data  // 검증된 데이터만 반환
}

// NetworkService: 디코딩 수행
return try await responseHandler.handleAsync(data: data, response: httpResponse)
    .decode(type: T.self, decoder: decoder)  // ❌ NetworkService가 디코딩
```

**개선 후 (올바른 책임 분리)**
```swift
// ResponseHandler: HTTP 검증 + 디코딩 통합
public func handle<T: Decodable>(data: Data, response: HTTPURLResponse) async throws(NetworkError) -> T {
    // 1. HTTP 상태 코드 검증
    let statusCode = response.statusCode
    guard (200..<300).contains(statusCode) else {
        let errorMessage = try? decoder.decode(BaseResponse<EmptyResponse>.self, from: data).message
        throw NetworkError.from(statusCode: statusCode, message: errorMessage)
    }

    // 2. 디코딩
    do {
        let decodedValue = try decoder.decode(T.self, from: data)
        return decodedValue
    } catch {
        throw NetworkError.decodingFailed(error)
    }
}

// NetworkService: 요청 생성 및 실행에만 집중
return try await handleResponse(data: data, response: httpResponse)  // 응답 처리 위임
```

## 👀 기타 더 이야기해볼 점
<!-- 있으면 작성하고 없으면 제목까지 완전히 지워주세요! -->
- 서버에서 아직 구현 안된 부분은 파일만 만들어 두었습니다 ~! 구현하실 때 한번 인덱스 번호 따라서 확인 부탁드려욥
- 원래 DTO 네이밍은 맨 앞에 Fetch, Delete, Update, Create를 붙여 지었는데 로그인, 로그아웃 같은 경우 애매해서 Request를 앞에 붙였습니다! 맘에 썩 들지 않는데 해결 방법을 잘 모르겠어서 ㅜㅜ 함께 이야기해보면 좋을 듯 합니닷

## 🔗 연결된 이슈
- Resolved: #16

youz2me committed:
[Refactor] #16 - NetworkError 네트워크 연결 유실 케이스 추가

youz2me committed:
[Refactor] #16 - endpoint를 path로 리팩토링

youz2me committed:
[Refactor] #16 - 번들 사용 시 network 모듈의 번들을 사용하도록 변경

youz2me committed:
[Refactor] #15 - ErrorResponse와 BaseResponse 통합

youz2me committed:
[Refactor] #16 - NetworkService를 Swift 6 async/await + typed throws로 변환

- Combine 제거 및 async/await 기반으로 전환
- typed throws(NetworkError) 적용으로 타입 안전성 강화
- ResponseHandler를 async/await 호환으로 변경
- Alamofire의 async API(serializingData) 사용
- 응답 처리 책임을 ResponseHandler로 일원화 (HTTP 검증 + 디코딩)

youz2me committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

youz2me committed:
[Feat] #16 - 범용 성능 측정 도구 PerformanceMonitor 추가

- 동기/비동기 작업 모두 측정 가능
- 소요 시간 및 메모리 사용량 측정
- 자동 측정(클로저) 및 수동 측정(start/end) 지원
- Thread-safe 구현
- DI 컨테이너 등록 가능한 구조

youz2me committed:
[Feat] #16 - NetworkMonitor 및 LoggingMonitor 추가

- NetworkMonitor 프로토콜: 네트워크 요청/응답 모니터링 인터페이스
- LoggingMonitor: 네트워크 로깅 구현체
  - 요청/응답 정보 출력 (URL, Method, Headers, Body)
  - JSON pretty print 지원
  - 에러 로깅 지원

youz2me committed:
[Refactor] #16 - NetworkService를 Swift 6 async/await + typed throws로 변환

- Combine 제거 및 async/await 기반으로 전환
- typed throws(NetworkError) 적용으로 타입 안전성 강화
- ResponseHandler를 async/await 호환으로 변경 (BaseResponse 사용)
- Alamofire async API(serializingData) 사용
- NetworkMonitor 통합으로 로깅 지원
- 응답 처리 책임을 ResponseHandler로 일원화

youz2me committed:
[Refactor] #16 - 네트워크 레이어 구조 개선 및 NetworkMonitor 제거

- LoggingMonitor를 NetworkHelper 폴더로 이동
- NetworkService에서 NetworkMonitor 의존성 제거
- MultipartData 제거 (미사용)
- 단일 책임 원칙에 따라 NetworkService 간소화

youz2me committed:
[Feat] #16 - HomeFeature DTO 구현

- FetchConcertMainSetlist: 대표 셋리스트 조회
- FetchHomeSectionList: 홈 화면 섹션 목록 조회
- FetchUserInterestConcert: 관심 콘서트 조회
- UpdateUserInterestConcert: 관심 콘서트 설정

CodingKeys를 통한 JSON 키 매핑 적용

youz2me committed:
[Feat] #16 - ConcertFeature DTO 구현

- FetchConcertInfo: 콘서트 정보 조회
- FetchConcertArtistInfo: 아티스트 정보 조회
- FetchConcertCultureList: 콘서트 문화 목록 조회
- FetchConcertKeyInfoList: 필수 정보 목록 조회
- FetchConcertMDList: MD 목록 조회
- FetchConcertSchedule: 일정 목록 조회
- FetchConcertSetlistList: 셋리스트 목록 조회

옵셔널 필드 처리 및 CodingKeys 매핑 적용

youz2me committed:
[Feat] #16 - SearchFeature DTO 구현

- FetchBannerList: 배너 목록 조회
- FetchConcertList: 콘서트 목록 조회 (페이지네이션)
- FetchFliterSearchResult: 필터 검색 결과 조회 (페이지네이션)
- FetchRecommendKeywordList: 추천 검색어 조회
- FetchSearchSectionList: 탐색 섹션 목록 조회

Cursor 기반 페이지네이션 구조 적용

youz2me committed:
[Feat] #16 - SetlistFeature DTO 구현

- FetchConcertSetlist: 셋리스트 상세 조회
- FetchSetlistSongList: 셋리스트 곡 목록 조회

typealias를 활용한 배열 타입 선언

youz2me committed:
[Feat] #16 - SongFeature DTO 구현

- FetchSongFanchant: 곡 응원법 및 포인트 조회
- FetchSongLyrics: 가사 정보 조회 (원어, 발음, 해석)

배열 데이터 구조로 가사 및 응원법 관리

youz2me committed:
[Feat] #16 - CommentFeature DTO 구현

- CreateCommentReport: 댓글 신고 생성
- CreateConcertComment: 댓글 작성
- FetchConcertCommentList: 댓글 목록 조회 (페이지네이션)

Request/Response 분리 및 Cursor 기반 페이지네이션

youz2me committed:
[Feat] #16 - OnboardingFeature DTO 구현

- CreateUser: 회원가입
- FetchUserInfo: 사용자 정보 조회
- RequestAppleLogin: Apple 로그인
- RequestKakaoLogin: Kakao 로그인
- UpdateToken: 토큰 재발급

인증 관련 Request/Response 구조 정립

youz2me committed:
[Feat] #16 - UserFeature DTO 구현

- DeleteUser: 회원 탈퇴
- RequestLogout: 로그아웃
- UpdateUserNickname: 닉네임 변경

사용자 관리 관련 Request/Response 구조 정립

youz2me committed:
[Refactor] #16 - LoggingMonitor 불필요한 init 제거

기본 init이 자동 생성되므로 명시적 선언 제거

JinUng41 commented on Livith-iOS/Projects/Network/Sources/NetworkHelper/LoggingMonitor.swift:
```diff
@@ -0,0 +1,60 @@
+
+import Foundation
+
+public final class LoggingMonitor {
```
구현상 `NetworkMonitor`를 따라야 하는 것이 아닌가 싶은데 까먹으신 걸까요?

youz2me replied:
헉. 이슈가 있어서 NetworkMonitor 채택을 없애놨는데 다시 붙이는 걸 까먹었습니다.
근데 말씀해주신 EventMonitor를 사용하는 식으로 바꿔볼 예정이라 NetworkMonitor는 삭제하려고 합니다!

JinUng41 commented on Livith-iOS/Projects/Network/Sources/NetworkMonitor.swift:
```diff
@@ -0,0 +1,19 @@
+
+import Foundation
+
+public protocol NetworkMonitor {
```
AF의 `EventMonitor`와의 차이는 무엇인가요?

youz2me replied:
음? ............................................................. 그 친구를 까맣게 잊고 있었네요!
리팩토링하도록 하겠습니다.

JinUng41 commented on Livith-iOS/Projects/Network/Sources/NetworkService.swift:
```diff
@@ -6,82 +6,90 @@
         responseHandler: ResponseHandlerProtocol = ResponseHandler(),
-        errorMapper: ErrorMapperProtocol = ErrorMapper()
+        errorMapper: ErrorMapperProtocol = ErrorMapper(),
+        interceptor: RequestInterceptor
```
`interceptor`는 필요없을 경우에도 대비하기 위해, 옵셔널로 주입되는 것이 좋을 것 같습니다.

그리고 앞서 구현하셨던 `NetworkMonitor`도 가져야 할 것 같네요.

AF의 `Session`을 주입하는 것과 `AF.request`를 실행하는 것의 차이가 뭘까요?

youz2me replied:
interceptor가 필요없는 경우가 있을까 싶어서 옵셔널 처리를 생각하지 않았는데 온보딩의 경우에는 필요가 없군요!
수정했습니닷.

> AF의 Session을 주입하는 것과 AF.request를 실행하는 것의 차이가 뭘까요?

이 부분은 잘 몰라서 공부를 좀 해봤는데 Session 객체 생성 시 인터셉터와 EventMonitor를 인자로 받기 때문에 커스터마이징이 가능하다는 점이 가장 큰 차이인 것 같습니닷. 테스트 때도 Mock Session을 생성할 수 있어서 용이하다는 점도 있을 것 같아요!

JinUng41 commented on Livith-iOS/Projects/Network/Sources/DTO/CommentFeature/CreateCommentReport.swift:
```diff
@@ -0,0 +1,37 @@
+
+import Foundation
+
+public extension DTO.Request {
+    struct CreateCommentReport: Encodable {
```
구조체 선언 시에도 `public`을 선언해야, 사용하는 다른 모듈에서 호출할 수 있지 않나요?

youz2me replied:
구조체에 public을 명시하면 컴파일 경고가 발생합니다!

JinUng41 commented on Livith-iOS/Projects/Core/Sources/PerformanceMonitor.swift:
성능 측정 유틸 코드를 구현하셨군요.
다양한 상황에서 유용하게 사용할 수 있을 것 같아요.

아이디어를 더 내보자면, 측정하려는 모듈에 대해서 테스트 모듈을 만들고 `XCTest`의 `measure`를 이용하는 것도 좋을 것 같아요.

youz2me replied:
그것도 좋은 방법이네요!! 리팩토링할 때 얼마나 개선됐는지 확인하기 위해서 만들어둔 코드인데 일단 임시로 사용하고 추후에 같이 의논해서 만들어봐도 좋을 것 같습니닷. 좋은 의견 감사합니다 ~!

JinUng41 reviewed (approved):
고생하셨습니다.
많은 DTO 객체 선언과 NetworkService 개선, 그리고 성능 측정 코드 구현까지~

DTO 선언과 관련하여, DTO을 기능의 Data 모듈에서 선언하는 것과 DTO를 네트워크 모듈에 선언하는 것에 대한 생각을 들어보고 싶어요.

youz2me committed:
[Fix] #16 - DTO 접근제어자 Public 명시하도록 수정

youz2me committed:
[Fix] #16 - gitignore 파일 클로드 관련 폴더 추가

youz2me committed:
[Fix] #16 - DTO 오타 수정

youz2me committed:
[Refactor] #16 - NetworkService에 Session 주입 및 EventMonitor 적용

- LoggingMonitor를 Alamofire EventMonitor로 변경하여 자동 모니터링 구현
- NetworkService에 Session 주입 방식 도입으로 테스트 용이성 및 확장성 개선
- AF.request 대신 session.request 사용으로 커스터마이징 가능
- Interceptor를 optional로 변경하여 인증 불필요 API 대응
- 사용하지 않는 NetworkMonitor 프로토콜 제거
